happi/audit.py

```diff
@@ -0,0 +1,424 @@
+            path_to_config = Client.find_config()
+
+            if path_to_config:
+                logger.info('Using Client cfg path %s ', path_to_config)
+                config = ConfigParser()
+                config.read(path_to_config)
+                try:
+                    database_path = config.get('DEFAULT', 'path')
+                except Exception as e:
+                    logger.error('Error when trying '
+                                 'to get database path %s', e)
+                    return
+                else:
+                    self.process_args(database_path, args)
+            else:
+                logger.error('Could not find the Happi Configuration file')
+                sys.exit(1)
+
+    def process_args(self, database_path, args):
+        """
+        Checks to see if a valid path to database was provided.
+        If --extras is provided, will call for check xtra attributes, otherwise
+        it will just proceed with parsing and validating the database call.
+        """
+        if self.validate_file(database_path):
+            logger.info('Using database file at %s ', database_path)
+            if args.extras:
+                self.check_extra_attributes(database_path)
+            else:
+                self.parse_database(database_path)
+        else:
+            logger.error('The database %s file path '
+                         'could not be validated', database_path)
+            sys.exit(1)
+
+    def validate_file(self, file_path):
+        """
+        Tests whether a path is a regular file
+
+        Parameters
+        -----------
+        file_path: str
+            File path to be validate
+
+        Returns
+        -------
+        bool
+            To indicate whether the file is a valid regular file
+        """
+        return os.path.isfile(file_path)
+
+    def validate_container(self, item):
+        """
+        Validates container definition
+        """
+        container = item.get('type')
+        device = item.get('name')
+        if container and container not in registry:
+            logger.error('Invalid device container: %s for device %s',
+                         container, device)
+            return self.report_code.INVALID
+        elif not container:
+            logger.error('No container provided for %s', device)
+            return self.report_code.MISSING
+        else:
+            # seems like the container has been validated
+            return self.report_code.SUCCESS
+
+    def validate_args(self, item):
+        """
+        Validates the args of an item
+        """
+        return [self.create_arg(item, arg) for arg in item.args]
+
+    def validate_kwargs(self, item):
+        """
+        Validates the kwargs of an item
+        """
+        return dict((key, self.create_arg(item, val))
+                    for key, val in item.kwargs.items())
+
+    def create_arg(self, item, arg):
+        """
+        Function borrowed from loader to create
```

Comment:
Please pull this function out in the loader so it can be reused rather than duplicating it here.

Reply:
Ok sounds good, I pulled this function out in 4fabfae, but still left it in the loader - is this what you meant?